File: src/main/java/me/pcasaes/hexoids/domain/service/GameLoopService.java
package me.pcasaes.hexoids.domain.service;

import me.pcasaes.hexoids.domain.config.Config;
import me.pcasaes.hexoids.domain.model.Game;

import java.util.Optional;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Used to add events into the game loop.
 */
public class GameLoopService {

    private static final GameLoopService INSTANCE = new GameLoopService();

    public static GameLoopService getInstance() {
        return INSTANCE;
    }

    private static final String NAME = "game-loop";
    private static final Logger LOGGER = Logger.getLogger(GameLoopService.class.getName());

    private final Optional<Runnable> fixedUpdateRunnable = Optional.of(this::fixedUpdate);


    private long nextFixedUpdateTime;

    private long getGameTime() {
        return Game.get().getClock().getTime();
    }

    private GameLoopService() {
        this.nextFixedUpdateTime = this.getGameTime();
    }

    private void fixedUpdate() {
        long timestamp = this.getGameTime();
        if (timestamp > this.nextFixedUpdateTime) {
            Game.get()
                    .fixedUpdate(timestamp);

            this.nextFixedUpdateTime = timestamp + Config.get().getUpdateFrequencyInMillis();
        }
    }

    public void accept(Runnable runnable) {
        try {
            runnable.run();
        } catch (RuntimeException ex) {
            LOGGER.log(Level.SEVERE, ex.getMessage(), ex);
        }
        fixedUpdate();
    }

    public String getName() {
        return NAME;
    }

    public Optional<Runnable> getFixedUpdateRunnable() {
        long timestamp = this.getGameTime();
        if (timestamp > nextFixedUpdateTime) {
            return fixedUpdateRunnable;
        }
        return Optional.empty();
    }

}

File: src/main/kotlin/me/pcasaes/hexoids/core/domain/service/GameLoopService.kt
package me.pcasaes.hexoids.core.domain.service

import me.pcasaes.hexoids.core.domain.config.Config
import me.pcasaes.hexoids.core.domain.model.Game
import java.util.logging.Level
import java.util.logging.Logger

/**
 * Used to add events into the game loop.
 */
object GameLoopService {

    private const val name = "game-loop"
    private val log: Logger = Logger.getLogger(GameLoopService::class.java.getName())


    private val fixedUpdateRunnable: Runnable = Runnable { this.fixedUpdate() }

    private var nextFixedUpdateTime: Long

    private fun gethGameTime(): Long {
        return Game.get().getClock().getTime()
    }

    init {
        this.nextFixedUpdateTime = this.gethGameTime()
    }

    private fun fixedUpdate() {
        val timestamp = this.gethGameTime()
        if (timestamp > this.nextFixedUpdateTime) {
            Game.get()
                .fixedUpdate(timestamp)

            this.nextFixedUpdateTime = timestamp + Config.getUpdateFrequencyInMillis()
        }
    }

    fun accept(runnable: Runnable) {
        try {
            runnable.run()
        } catch (ex: RuntimeException) {
            log.log(Level.SEVERE, ex.message, ex)
        }
        fixedUpdate()
    }

    fun getName(): String {
        return name
    }

    fun getFixedUpdateRunnable(): Runnable? {
        val timestamp = this.gethGameTime()
        return if (timestamp > nextFixedUpdateTime) {
            fixedUpdateRunnable
        } else {
            null
        }
    }


}
